### DISCUSSION TITLE
converting from tiff to HEIC changes colors significantly

### DISCUSSION THREAD

**Discussion by GabZach** [ImageMagick/ImageMagick#1947] [General]
I am trying to (bulk) convert tiff images to HEIC format.
However, when I try to do that using `convert in.tiff out.heic` , I get significantly different colors (duller).
I have ImageMagick 6.9.10-90 Q16 x86_64 2020-03-16, installed via MacPorts.
I do the conversion on macOS Catalina.

When I do the conversion using Preview on my Mac, I get images indistinguishable with the naked eye.

BTW: conversion from tiff to jpeg is better, but also not quite perfect: there is a slight difference in colors.

Here is the relevant output of `identify -list Format | grep -e HEIC -e TIFF -e JPG`:
```
     HEIC* HEIC      rw-   Apple High efficiency Image Format (1.6.2)
      JPG* JPEG      rw-   Joint Photographic Experts Group JFIF format (libjpeg 90)
     PTIF* TIFF      rw+   Pyramid encoded TIFF
     TIFF* TIFF      rw+   Tagged Image File Format (LIBTIFF, Version 4.1.0)
   TIFF64* TIFF      rw+   Tagged Image File Format (64-bit) (LIBTIFF, Version 4.1.0)
```

Here you can see the original I used for testing,
while the HEIC and the JPG images are the results from the convert command:

   https://owncloud.informatik.uni-bremen.de/index.php/s/JQgEG8BDEfAAkKB

Any ideas what I can do?

**Comment by fmw42** [comment]
Do your Apple converted HEIC image have any profiles? Is there a specific profile for HEIC?

**Comment by urban-warrior** [comment]
Try a compare:

```
compare -metric rmse \( Aiken-Walker,\ Cotton\ Pickers,\ 1885.heic -colorspace sRGB \) Aiken-Walker,\ Cotton\ Pickers,\ 1885.tiff null:
3.0619 (0.012008) 
```
That's a low distortion suggesting the conversion was accurate.

**Reply by GabZach** [reply]
I get 786.497 (0.0120012), and a pretty much red output diff image.

**Reply by GabZach** [reply]
But, actually, I don't see what I can learn from the `compare` , since I can see the color shifts in the output image directly on the monitor - they are pretty blatant.

**Reply by fmw42** [reply]
It tells you if the image data is identical. The difference you see can come from the data or the monitor due to different profiles.

**Reply by GabZach** [reply]
You mean it could happen that the HEIC image just "looks" different when rendered on the screen, but in the file, everything is correct?
This is what I get when executing your `compare` command:
![diff](https://user-images.githubusercontent.com/11965698/80460454-25944480-8934-11ea-8af4-d842951deb80.png)

**Comment by GabZach** [comment]
Sorry, I forgot to mention the profiles. Yes, the tiff image has the profile "Adobe RGB (1998)". 
The jpg output of `convert` also has the profile "Adobe RGB (1998)". Both jpg and tiff have the same color depth (8) and color model (RGB).
Still , the jpeg output of `convert` looks slightly different than the input tiff.

Now, the HEIC output of `convert` has the profile "sRGB IEC61966-2.1" - why would `convert` do that.
(Sure enough, the HEIC output of Preview has the profile "Adobe RGB (1998)".)

**Reply by fmw42** [reply]
Make sure you apply to AdobeRGB profile to your HEIC output of the convert command.

**Reply by GabZach** [reply]
How do I do that without knowing the profile of the input image? (I want to bulk convert)

**Reply by GabZach** [reply]
Or does the profile of the input image not matter? Should I apply the AdobeRGB profile to the output , no matter what profile the input has?

**Reply by fmw42** [reply]
You can apply the same profile if you want. The default is to take the input profile if it has one; otherwise, I believe Imagemgick will apply sRGB profile. If you script a loop over all your input files, you can extract the profile from the input in one command line and then decide if you want that profile or to apply Adobe RGB or sRGB in a second command line.

**Reply by GabZach** [reply]
If the default for `convert` is to take the profile of the input image (which is what I'd want), then why does the HEIC file have sRGB as profile, even though the input file has Adobe RGB?

**Comment by snibgo** [comment]
IM's normal action is to propagate any embedded profile from the first input file to the output file. So, for example,
`convert in.png out.tiff`
... out.tiff will contain the same pixel values as in.png, and the same embedded profile (if any).

When writing a HEIF, if IM is copying the pixels values, and discarding an embedded profile, and substituting a sRGB profile, this seems like bad behaviour. It is bad because the embedded profile no longer describes what the pixels should look like, so image viewers will display the wrong colours.

That bad behaviour may be within IM code or libheif library code. Or perhaps there is a switch within the library that IM should set.

A developer may comment.

**Reply by Jmuccigr** [reply]
Perhaps encountering this myself as my tiff -> heic results in heic files with sRGB profiles, even when I extract the "TrueColor sRGB" profile from the tiff first (to file "test.icc") and apply it to the output via `-profile test.icc`.
